src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttributes.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class TagHelperAttributes<TAttributeValue> : IList<TagHelperAttribute<TAttributeValue>>
+        where TAttributeValue : class
+    {
+        private readonly IList<TagHelperAttribute<TAttributeValue>> _attributes;
+
+        public TagHelperAttributes()
+        {
+            _attributes = new List<TagHelperAttribute<TAttributeValue>>();
+        }
+
+        public int Count => _attributes.Count;
+
+        public bool IsReadOnly => false;
+
+        public IEnumerable<string> Keys => _attributes.Select(attribute => attribute.Key);
+
+        public IEnumerable<TAttributeValue> Values => _attributes.Select(attribute => attribute.Value);
+
+        public TagHelperAttribute<TAttributeValue> this[[NotNull] string key] =>
+            _attributes.LastOrDefault(attribute => MatchesAttribute(key, attribute));
```

Comment:
in general this class looks inconsistent with everything else in Razor, less readable, and like @NTaylorMullen had fun with the new syntax.  was it an obfuscation exercise -- how much can you cram on a line or two?